Issue #41 - Remove keys

CodyGramlich opened the pull request:
Resolves issue [#41](https://github.com/yeg-relief/youcanbenefit/issues/41).

Main changes to look at:
-backend/src/modules/key/key.service.ts
-frontend/src/app/admin/screener/screener-toolbar/screener-toolbar.component.ts

Once I got it working without keys, I removed the key object from the frontend in many places, so a lot of the changes are due to that.

Notes:
- Since you cannot remove mappings in elasticsearch ([see here](https://www.elastic.co/guide/en/elasticsearch/reference/6.4/indices-delete-mapping.html)), I chose to back up the queries, delete the index, and upload the queries and the new mappings every time save screener is clicked.
- If a question belongs to a query and then that question is removed from the screener, I made it so that the conditions with that question are removed from the queries. If there are no conditions left in a query, then that query is deleted. This all happens when re-indexing.
- Queries in elasticsearch now contain a questionTexts object under meta to map the question ID to question text
- I replaced the key object
```
export interface Key {
  name: string;
  type: 'boolean' | 'integer' | 'number' | 'invalid' | '';
}
```
with
```
export interface QuestionKey {
    text: string;
    id: string;
    type: 'boolean' | 'integer' | 'number' | 'invalid' | '';
}
```
I was going to call this new data type Question, but there is already a type called Question and I didn't want to confuse the two. I went with QuestionKey to also distnguish it from a key. I wasn't sure of a good name for the object. This object type is used on both the frontend and the backend.
- Question texts appear for conditions instead of the key name
- For the most part, key.name has been replaced with questionKey.id. In some cases it has been replaced with questionKey.text, such as when displaying the question texts.
- Multiselect options now have an ID associated with them.
- Keys section has been removed.
- Key filter has been removed.
- There is no longer a drop down menu to select a key for questions and for multiselect options.
- Question_2 interface has been removed.
- The Key attribute of the Question interface has been removed.
- The NumberSelect type of question (select one from the following) has been disabled. I could update it so that it handles this, but for now I left it disabled.
- I haven't changed the name of KeyService or KeyModule in the backend.
- Many tests need to be updated.
- I pulled from release/10 to get Steven's work with mine and resolved the merge conflicts.
- To test it out, you must use Init and then create programs, questions, queries. Uploading an old backup will not work because the key names need to be replaced with question ID's, along with some other stuff. I can convert our current backup to this new format later.

Any feedback would be great.

slmyers commented:
I intend to review this before the weekend.

slmyers commented on backend/src/modules/query/EsQuery.dto.ts:
```diff
@@ -53,12 +53,19 @@ type Condition =
 
-type Meta = {[key: string]: string}
+
+// type Meta = {[key: string]: string} | {[key:string]: {[key: string]: string}}
```
Please delete this comment.

slmyers commented on frontend/src/app/admin/data.service.ts:
```diff
@@ -12,23 +11,8 @@ import { environment } from '../../environments/environment'
 
 @Injectable()
 export class DataService {
-    private keys$: Observable<Key[]>;
```
I believe this class is now unused. Please delete it and the import in `admin\admin.module.ts`.

slmyers commented on frontend/src/app/admin/screener/question-edit/mult-select-questions/mult-select-questions.component.html:
```diff
@@ -1,14 +1,7 @@
-        </mat-select>
-    </div>
-    <button mat-button [style.background-color]="'darkorange'" (click)="commit()">Save</button>
+    <button mat-button [style.background-color]="'darkorange'" (click)="commit()" [disabled]="!form.valid ? true : null">Save</button>
```
you can avoid the ternary for the `disabled` prop

```typescript
[disabled]="!form.valid"
```

slmyers commented on frontend/src/app/admin/screener/question-edit/question-edit.component.html:
```diff
@@ -36,7 +25,7 @@
         </mat-select>
 
-        <section [formGroup]="optionForm" *ngIf="controlType === 'NumberSelect'">
+        <!-- <section [formGroup]="optionForm" *ngIf="controlType === 'NumberSelect'">
```
Please remove the commented markup.

slmyers commented on frontend/src/app/admin/screener/question-edit/question-edit.component.spec.ts:
```diff
@@ -34,11 +29,8 @@ const screenerState: fromScreener.State  = {
-  ],
-  created: 0
+  created: 0,
+  DATA: "TEST"
```
please remove test value

slmyers commented on frontend/src/app/admin/screener/question-edit/question-edit.component.ts:
```diff
@@ -35,14 +35,13 @@ import { Animations } from '../../../shared/animations'
     readonly CONTROL_TYPE_VALUES = [
         { value: 'NumberInput', display: 'number input' },
-        { value: 'NumberSelect', display: 'select' },
+        // { value: 'NumberSelect', display: 'select' },
```
please remove commented code in this file.

slmyers commented on frontend/src/app/admin/screener/screener-overview/screener-overview.component.spec.ts:
```diff
@@ -42,11 +41,8 @@ const screenerState: fromScreener.State  = {
-  ],
-  created: 0
+  created: 0,
+  DATA: 'TEST'
```
please remove test value

slmyers commented on frontend/src/app/admin/screener/screener-toolbar/screener-toolbar.component.spec.ts:
```diff
@@ -38,11 +32,8 @@ const screenerState: fromScreener.State  = {
-  ],
-  created: 0
+  created: 0,
+  DATA: 'TEST'
```
please remove test value

CodyGramlich commented:
Hi Steven,

Thanks for reviewing my pull request. I'll remove some of those comments and that data service. I also pushed a few more changes:
- Replaced the key module with a question module that has two functions: updateQuestions() and getQuestions()
- Moved the following functions from the old key service to EsQuery.service.ts: backup queries, update queries, delete query index, create query index, put mappings, upload queries
- Renamed the object QuestionKey to Question and the object for the screener questions from Question to ScreenerQuestion
- Combined the backend update questions function and the update screener function so that both are called at protected/screener/ and only one request is sent when the save screener button is clicked
- Added a snackbar for the save screener response

As for the error that you ran into, I think the error only comes up when the current database is uploaded. I will have to transform the current database to this new format.

j-rewerts commented:
Woohoo community!! 🎉 

@CodyGramlich Once you've got most of our comments addressed, please make sure the seed script works with our integration tests, and make sure you have a transformed backup.json that we can play around with in the test environment, just to prove we can rebuild ES entirely from within YCB.

Great job!

CodyGramlich commented:
I shared the transformed backup on slack.

We will need to do some adjusting of the query display to accommodate for question texts. We also could remove the Type column as the type can be implied from the question text or value.

slmyers commented:
i'm working on dry-run queries (#11) so I'll see if I can address the query display.
